lib/linked-list.js

```diff
@@ -9,105 +9,198 @@ class Node {
     index count starts at 0
     returns nil if there are fewer nodes in the linked list than the index value
-    Time Complexity: ?
-    Space Complexity: ?
+    Time Complexity: O(N)
+    Space Complexity: O(1)
      */
     getAtIndex(index) {
-        throw new Error("This method hasn't been implemented yet!");
+        // if (index >= this.length || !this.#head) { return null }
+        // if (index === 0) { return this.#head.value }
+        // Loop through index
+        let i = 0;
+        let current = this.#head;
+        while (current) {
+            if (i === index) {
+                return current.value
+            } else {
+                i++ // increment index value
+                current = current.next // increment value of current
+            }
+        }
+
+        return null
     }
 
     /*
     method that returns the value of the last node in the linked list
     returns nil if the linked list is empty
-    Time Complexity: ?
-    Space Complexity: ?
+    Time Complexity: O(1)
+    Space Complexity: O(1)
     */
     getLast() {
-        throw new Error("This method hasn't been implemented yet!");
+        return this.tail.value
     }
 
     /*
     method that inserts a given value as a new last node in the linked list
-    Time Complexity: ?
-    Space Complexity: ?
+    Time Complexity: O(1)
+    Space Complexity: O(1)
     */
     addLast(value) {
-        throw new Error("This method hasn't been implemented yet!");
+        // If tail, store value
+        // Else, set new node to tail
+        if (!this.tail) {
+            this.addFirst(value);
+        } else {
+            const newNode = new Node(value);
+            this.tail.next = newNode; // set current tail next node to new node
+            this.tail = newNode; // set current tail to new node
+            this.#length++
+        }
     }
 
     /*
     method to return the max value in the linked list
     returns the data value and not the node
-    Time Complexity: ?
-    Space Complexity: ?
+    Time Complexity: O(N)
+    Space Complexity: O(1)
     */
     findMax() {
-        throw new Error("This method hasn't been implemented yet!");
+        let maxVal;
+
+        if (!this.#head) { maxVal = null }
+        else {
+            maxVal = 0;
+            let current = this.#head; // Start at head node
+            // Traverse the list finding the max
+            // Compare current value to current max
+            while (current) { // Current is not null, should reach tail
+                if (current.value > maxVal) { maxVal = current.value } // If node value > maxVal, update maxVal
+                current = current.next; // Move to the next node
+            }
+        }
+        return maxVal
     }
 
     /*
     method to delete the first node found with specified value
-    Time Complexity: ?
-    Space Complexity: ?
+    Time Complexity: O(N)
+    Space Complexity: O(1)
     */
     delete(value) {
```

Comment:
👍  Nicely done.  I do want to note that this becomes a bit easier with a doubly linked list (you don't need the previous reference).